fixql-core/src/main/scala/com/iterable/graphql/FromGraphQLJava.scala

```diff
@@ -0,0 +1,202 @@
+import graphql.schema.idl.SchemaGenerator
+import graphql.schema.GraphQLSchema
+import graphql.schema.idl.RuntimeWiring
+import graphql.schema.idl.SchemaParser
+import io.circe.Json
+import io.circe.JsonNumber
+import io.circe.JsonObject
+import play.api.libs.json.{Json => PlayJson}
+import play.api.libs.json.JsArray
+import play.api.libs.json.JsBoolean
+import play.api.libs.json.JsNull
+import play.api.libs.json.JsNumber
+import play.api.libs.json.JsString
+import play.api.libs.json.JsValue
+
+import scala.concurrent.Future
+import scala.collection.JavaConverters.iterableAsScalaIterableConverter
+import scala.collection.JavaConverters.mapAsScalaMapConverter
+import scala.collection.JavaConverters.mapAsJavaMapConverter
+import scala.util.Try
+
+/**
+  * Document and schema parsing using the GraphQL Java library.
+  */
+object FromGraphQLJava {
+
+  /** Parse and validate documents (queries and mutations) using the GraphQL Java library.
+    * This requires a GraphQL Java GraphQLSchema. Returns our Query type.
+    *
+    * Overall flow looks as follows:
+    * - Build a GraphQL Java GraphQLSchema whether from SDL or programmatically.
+    * - Use GraphQL Java to parse the document (i.e. query) into Java AST
+    * - Use GraphQL Java to resolve variables etc.
+    * - Convert to Scala AST by traversing Java using selection-set peeking. This is what the code below does.
+    * - Compile Scala AST into DBIO
+    * - Run DBIO
+    *
+    * This method uses "field selection sets" from GraphQL-Java to extract a Field tree for the query:
+    * https://www.graphql-java.com/documentation/v12/fieldselection/
+    */
+  def parseAndValidateQuery(graphQLSchema: GraphQLSchema, query: String, variables: Json): Try[Query[Field.Fixed]] = {
+    val extractExecutionStrategy = new ContextExtractingExecutionStrategy
+    Try {
+      val (context: ExecutionContext, fieldSubSelection: FieldSubSelection) = graphql.nextgen.GraphQL.newGraphQL(graphQLSchema)
+        .executionStrategy(extractExecutionStrategy)
+        .build()
+        .execute(ExecutionInput.newExecutionInput(query)
+          .variables(toJavaValues(variables).asInstanceOf[Map[String, AnyRef]].asJava)
+          .build())
+        .getData[java.util.LinkedHashMap[String, Any]]
+        .get(null) // the map will have a value with key = null
+
+      val valuesResolver = new ValuesResolver
+      val topLevelFields: Seq[Field.Fixed] =
+        fieldSubSelection.getMergedSelectionSet.getSubFieldsList.asScala.map { mergedField =>
+          // See DataFetchingSelectionSetImpl.traverseFields
+          val fieldDef = Introspection.getFieldDef(graphQLSchema, graphQLSchema.getQueryType, mergedField.getName)
+          val unwrappedType = GraphQLTypeUtil.unwrapAll(fieldDef.getType)
+          // See ValueFetcher.fetchValue() or other callers of DataFetchingFieldSelectionSetImpl.newCollector()
+          //val codeRegistry: GraphQLCodeRegistry = executionContext.getGraphQLSchema.getCodeRegistry
```

Comment:
Are you planning to do something with the commented out code?

Reply:
i'll remove them